### PR TITLE
[Core] Rename ClientFactoryFn to CoreWorkerClientFactoryFn

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?

Make it clear that it's about core worker client, not raylet client or gcs client.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by dayshah — APPROVED** [review]

File: src/ray/rpc/worker/core_worker_client_pool.h [hunk 1]
```diff
@@ -76,19 +73,19 @@ class CoreWorkerClientPool {
...
+  CoreWorkerClientFactoryFn client_factory_;
 
   absl::Mutex mu_;
 
   struct CoreWorkerClientEntry {
    public:
     CoreWorkerClientEntry() {}
     CoreWorkerClientEntry(ray::WorkerID worker_id,
-                          shared_ptr<CoreWorkerClientInterface> core_worker_client)
+                          std::shared_ptr<CoreWorkerClientInterface> core_worker_client)
```
Comment (dentiny) [inline src/ray/rpc/worker/core_worker_client_pool.h:84]: Can be moved?

File: src/ray/rpc/worker/core_worker_client_pool.h [hunk 2]
```diff
@@ -35,13 +32,13 @@ class CoreWorkerClientPool {
       : client_factory_(defaultClientFactory(ccm)){};
 
   /// Creates a CoreWorkerClientPool by a given connection function.
-  CoreWorkerClientPool(ClientFactoryFn client_factory)
+  CoreWorkerClientPool(CoreWorkerClientFactoryFn client_factory)
```
Comment (dentiny) [inline src/ray/rpc/worker/core_worker_client_pool.h:35]: `explicit` and move semantics

File: src/ray/rpc/worker/core_worker_client_pool.h [hunk 3]
```diff
@@ -20,9 +20,6 @@
 #include "ray/common/id.h"
 #include "ray/rpc/worker/core_worker_client.h"
 
-using absl::optional;
```
Comment (dentiny) [inline src/ray/rpc/worker/core_worker_client_pool.h:23]: Oh my....

**Review by dentiny — APPROVED** [review]
Only two comments on C++

File: src/ray/rpc/worker/core_worker_client_pool.h (hunk 1 above)
Comment (jjyao) [reply src/ray/rpc/worker/core_worker_client_pool.h:84]: Will fix in my next PR

File: src/ray/rpc/worker/core_worker_client_pool.h (hunk 2 above)
Comment (jjyao) [reply src/ray/rpc/worker/core_worker_client_pool.h:35]: Will fix in my next PR

**Pull request merged**